Add Suggest component

andreadelrio opened the pull request:
Fixes https://github.com/elastic/eui/issues/1938
Fixes https://github.com/elastic/eui/pull/2270

### Summary

This PR adds the `EuiSuggest` component and the corresponding docs. 

This component provides a way to display suggestions using a text field. Behind the scenes it uses `EuiSuggestItem` to display each suggestion following a very similar format to the one currently present in Kibana.

The consumer is to handle what the current status of `EuiSuggest` is. The supported status are `notYetSaved`, `saved`,  `noNewChanges` and `isLoading`. Custom tooltip content is supported for these.

`EuiSuggest` can also have elements prefixed and/or appended to it. For this, there's the `prefix` and `append` props which are both `PropTypes.node`.

![image](https://user-images.githubusercontent.com/4016496/64276420-c85b0700-cefc-11e9-9d22-3b80209f9420.png)
![image](https://user-images.githubusercontent.com/4016496/64277297-dad64000-cefe-11e9-9078-42044ff4a02d.png)


### Checklist

- [x] Checked in **dark mode**
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
- [x] Props have proper **autodocs**
- [x] Added **documentation** examples
- [x] Added or updated **jest tests**
- [x] Checked for **breaking changes** and labeled appropriately
~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [X] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

cchaos commented:
@andreadelrio Is this ready for review?

andreadelrio commented:
@cchaos I realized I need to clean some things up still. I'll let you know when it's ready. Thanks

cchaos commented on src-docs/src/views/suggest/_global_filter_group.scss:
```diff
@@ -17,12 +17,3 @@
   transition: height $euiAnimSpeedNormal $euiAnimSlightResistance;
 }
-
-.globalFilterGroup__filterFlexItem {
```
Looks like there was a bad merge here. This line and below shouldn't be removed.

cchaos commented on src-docs/src/views/suggest/_global_filter_item.scss:
```diff
@@ -26,7 +26,5 @@
     left: 0;
     width: $euiSizeXS;
     background-color: $euiColorVis0;
-    border-top-left-radius: $euiBorderRadius / 2;
```
Another bad merge here. This line and below shouldn't be removed.

cchaos commented on src-docs/src/views/suggest/global_filter_bar.js:
```diff
@@ -13,10 +13,7 @@ export const GlobalFilterBar = ({ filters, className, ...rest }) => {
-        <EuiFlexItem
-          key={filter.id}
-          grow={false}
-          className="globalFilterBar__flexItem">
```
Another bad merge here. This line shouldn't be removed.

cchaos commented on src-docs/src/views/suggest/global_filter_bar.js:
```diff
@@ -26,10 +23,7 @@ export const GlobalFilterBar = ({ filters, className, ...rest }) => {
-        <EuiFlexItem
-          key={filter.id}
-          grow={false}
-          className="globalFilterBar__flexItem">
```
Another bad merge here. This line shouldn't be removed.

cchaos commented on src-docs/src/views/suggest/global_filter_item.js:
```diff
@@ -188,8 +188,7 @@ export class GlobalFilterItem extends Component {
         button={button}
         anchorPosition="downCenter"
-        panelPaddingSize="none"
-        display="block">
```
Another bad merge here. This line shouldn't be removed.

cchaos commented on src-docs/src/views/suggest/saved_queries.js:
```diff
@@ -0,0 +1,267 @@
+
+    return (
+      <div className="savedQueriesInput">
+        <EuiFlexGroup>
```
This wrapping FlexGroup & Item isn't necessary because there's only a single child.

cchaos commented on src-docs/src/views/suggest/saved_queries.js:
```diff
@@ -0,0 +1,267 @@
+      <div className="savedQueriesInput">
+        <EuiFlexGroup>
+          <EuiFlexItem>
+            <EuiSelect
```
I think this would be nice to see all the options without needing to click a select box first. Consider making this a radio group.

cchaos commented on src-docs/src/views/suggest/saved_queries.js:
```diff
@@ -0,0 +1,267 @@
+              options={this.options}
+              value={this.state.status}
+              onChange={this.onSelectChange.bind(this)}
+              aria-label="Use aria labels when no actual label is in use"
```
You'll need an actual aria-label that describes/labels this input.

cchaos commented on src/components/suggest_item/_suggest_item.scss:
```diff
@@ -5,6 +5,18 @@
   font-size: $euiFontSizeXS;
   white-space: nowrap;
 
+  &:hover {
```
Hover states should only be added if the suggest item has an onClick or some other interaction. And then don't forget about focus states too.

cchaos replied:
Looks like you've got duplicates of this file. There's also a `_suggest_item.scss` in the folder above. I think you've been working in the other file, but the location of it should be here.

cchaos commented on src/components/suggest/suggest_input.js:
```diff
@@ -0,0 +1,145 @@
+    const statusMap = {
+      notYetSaved: {
+        icon: 'dot',
+        color: '#DD0A73',
```
This is the same color as our `accent`.

cchaos commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,92 @@
+    this.state = {
+      value: '',
+      status: 'notYetSaved',
+      isPopoverOpen: false,
```
All the popover stuff can be removed from this file.

andreadelrio commented:
![tootlip](https://user-images.githubusercontent.com/4016496/64200193-d0ee0780-ce40-11e9-8c2e-eaf5a426b557.gif)
@snide Question about tooltip behaviour. I'm currently using the copy you had in the mockups but as Caroline points out above the tooltips are not clickable so the copy doesn't work. I think we should change the copy instead of making the tooltip clickable. Thoughts?

@cchaos @snide Related question, we're setting the input states for the consumer as  `notYetSaved`, `saved`,  `noNewChanges` and `isLoading`. Should the tooltip text be customizable? If so would that be four more props (or three if `isLoading` doesn't get one) for `EuiSuggest`? or maybe 1 array?

cchaos commented:
> I think we should change the copy instead of making the tooltip clickable. Thoughts?

I agree, the copy should change. 

> Should the tooltip text be customizable?

I think that sounds easy enough. However, you don't need them to pass a tooltip content for each state, you can just allow them to override the current state's content. For instance, you are currently leaving it up to the consumer to pass down what the current state of the Suggest should be. Like `<Suggest status="notYetSaved" />`. So if they want to change the tooltip too they can do `<Suggest status="notYetSaved" tooltipContent="My custom content for the current state" />`. Since they'll know when the state changes because **they** are the ones tracking that, they'll just change those props at the same time like:

```
<Suggest
  status={this.state.saved ? 'saved' ? 'notYetSaved'}
  tooltipContent={this.state.saved ? 'My saved tooltip' ? 'My notYetSaved tooltip'}
/>
```

cchaos commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,92 @@
+   * Status of the current query 'notYetSaved', 'saved', 'noNewChanges' or 'isLoading'.
+   */
+  status: PropTypes.oneOf([
+    'notYetSaved',
```
I'm just realizing that this should probably be `unsaved` to be more succinct.

cchaos commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,92 @@
+  status: PropTypes.oneOf([
+    'notYetSaved',
+    'saved',
+    'noNewChanges',
```
And this one could then probably be `unchanged`.

andreadelrio commented:
@cchaos I think I've addressed most of your comments. Still not 100% happy with the behavior of the input on the Saved Queries example (it's a bit jumpy). Additionally, I added a handler for a click on a suggestion.

cchaos commented on src-docs/src/views/suggest/saved_queries.js:
```diff
@@ -0,0 +1,165 @@
+              sendInputValue={this.getInputValue.bind(this)}
+            />
+          </EuiFlexItem>
+          <EuiFlexItem grow={false} className="datepicker">
```
These class names are probably too generic. I was using `global` for the filters, you can probably do the same.

cchaos commented on src/components/suggest/suggest_input.js:
```diff
@@ -0,0 +1,140 @@
+  /**
+   * Element to be prepended to the input bar.
+   */
+  prefix: PropTypes.node,
```
Since you're not doing any manipulation on this prop and your putting `...rest` on the `EuiFieldText` component, I would remove this as a named prop and just allow consumers to pass it as `prepend`. 

Just like you're not explicitly stating that this component can be `disabled` or `readOnly` but if they added those props to EuiSuggest, it would get passed all the way to the field control and work properly.

cchaos commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,92 @@
+  /**
+   * Element to be prepended to the input bar (e.g. 'KQL').
+   */
+  prefix: PropTypes.node,
```
That would also mean that you can get rid of this named prop as well.

cchaos commented on src-docs/src/views/suggest/saved_queries.js:
```diff
@@ -0,0 +1,165 @@
+              status={this.state.status}
+              onFocus={this.onFieldFocus.bind(this)}
+              onBlur={this.onFieldBlur.bind(this)}
+              prefix={<HashtagPopover value={this.state.value} />}
```
And then this prop just gets renamed to `prepend`
```suggestion
              prepend={<HashtagPopover value={this.state.value} />}
```

cchaos commented on src-docs/src/views/suggest/suggest_example.js:
```diff
@@ -28,6 +38,27 @@ const suggestItemSnippet = [
 
+const suggestSnippet = [
+  `<EuiSuggest
+    status={this.state.status}
```
Total nit, but from this line down the snippet should be indented one tab less.

<img src="https://d.pr/free/i/SSTR1N+" width="50%" />

cchaos commented on src-docs/src/views/suggest/suggest_example.js:
```diff
@@ -45,10 +76,17 @@ export const SuggestExample = {
         <div>
           <p>
-            <EuiCode>EuiSuggest</EuiCode> description goes here.
+            <EuiCode>EuiSuggest</EuiCode> is a text field component to use
```
```suggestion
            <EuiCode>EuiSuggest</EuiCode> is a text field component used to
```

cchaos commented on src-docs/src/views/suggest/suggest_example.js:
```diff
@@ -67,15 +105,43 @@ export const SuggestExample = {
+        <div>
+          <EuiCallOut color="warning" title="Demo of visual pattern only">
+            <p>
+              This documents a <strong>visual</strong> pattern for the eventual
```
```suggestion
              This documents a <strong>visual</strong> pattern for
```

cchaos commented on src-docs/src/views/suggest/suggest_example.js:
```diff
@@ -67,15 +105,43 @@ export const SuggestExample = {
+          <EuiCallOut color="warning" title="Demo of visual pattern only">
+            <p>
+              This documents a <strong>visual</strong> pattern for the eventual
+              replacement of Kibana&apos;s global query and filter bars. The
```
```suggestion
              Kibana&apos;s global query and filter bars. The
```

cchaos commented on src/components/suggest/_suggest_input.scss:
```diff
@@ -0,0 +1,18 @@
+  }
+
+  .euiInputPopover {
+    max-width: 100%;
```
EuiInputPopover accepts a `fullWidth` prop that will apply this same property. I'd recommend using that instead.

cchaos commented on src-docs/src/views/suggest/suggest.js:
```diff
@@ -1,3 +1,112 @@
+          tooltipContent={this.state.tooltipContent}
+          sendInputValue={this.getInputValue.bind(this)}
+          onItemClick={this.onItemClick.bind(this)}
+          suggestions={sampleItems}
```
I'd suggest adding placeholder text to instruct the viewer of the docs to start typing to see the suggest component work.

cchaos commented on src/components/suggest/suggest_item.js:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  return (
+    <div onClick={onClick ? onClick : undefined} className={classes} {...rest}>
```
While technically div's can have `onClick`s, we've been steering towards making these `button` tags if `onClick` exists.

Similar to how EuiCard handles it:
https://github.com/elastic/eui/blob/c0b81336acd09f8186d795bb8630255e64493b47/src/components/card/card.js#L109-L114

andreadelrio replied:
![image](https://user-images.githubusercontent.com/4016496/64376415-09c2e380-cfdd-11e9-88c6-c6f52ec52bc5.png)
@cchaos  Using `button` breaks the styling. Any ideas why this could be happening?

cchaos replied:
It's probably a text-align issue. All browsers inherently give `<button>` tags `text-align: center`. You'll need to override that with `text-align: left`.

cchaos commented on src/components/suggest/suggest_input.js:
```diff
@@ -0,0 +1,140 @@
+    const statusElement = status !== 'isLoading' && (
+      <EuiToolTip
+        position="left"
+        content={tooltipContent}
```
I think you can still supply some good defaults in the component itself for the tooltip instead of forcing that on the consumer. Probably just add a few to the `statusMap` but allow `tooltipContent` to override.
```suggestion
        content={tooltipContent || statusMap[status].tooltip}
```

The `||` will evaluate the first option first and if it is false, move to the next option.

andreadelrio commented:
@cchaos managed to fix the styling when using `button` and made the other changes too. Component feels a bit cleaner now.

cchaos commented on src-docs/src/views/suggest/saved_queries.js:
```diff
@@ -0,0 +1,165 @@
+      <div className="savedQueriesInput">
+        <EuiFlexGroup
+          gutterSize="s"
+          className={this.state.hideDatepicker ? 'hideDatepicker' : ''}>
```
This className should update as well to something more specific.

cchaos commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,88 @@
+      onItemClick,
+      sendInputValue,
+      status,
+      prefix,
```
Don't forget to remove `prefix` from here

cchaos commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,88 @@
+    const suggestInput = (
+      <EuiSuggestInput
+        status={status}
+        prefix={prefix}
```
And here

cchaos commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,88 @@
+  /**
+   * Status of the current query 'notYetSaved', 'saved', 'unchanged' or 'isLoading'.
+   */
+  status: PropTypes.oneOf(['unsaved', 'saved', 'unchanged', 'isLoading']),
```
Sorry, I know I keep changing the names of these, but it `isLoading` feels like the odd one out. Maybe change to just `loading`?

cchaos commented on src/components/suggest/_suggest_item.scss:
```diff
@@ -5,21 +5,38 @@
+      cursor: pointer;
+      background-color: $euiColorLightestShade;
+
+      @each $name, $color in $euiSuggestItemColors {
```
Because of the additional specificity of `.euiSuggestItem-isClickable:hover`, you actually shouldn't need to loop through all the type colors to apply the same text color. You can just do the following and it will override the color you set below in the loop.:

```scss
.euiSuggestItem__type {
  color: $euiColorDarkestShade;
}
```

cchaos commented on src/components/suggest/suggest_input.js:
```diff
@@ -0,0 +1,137 @@
+      saved: {
+        icon: 'checkInCircleFilled',
+        color: 'secondary',
+        tooltip: 'Query successfully saved.',
```
Query is a bit too specific here, I'd suggest just making this very succinct
```suggestion
        tooltip: 'Saved',
```

cchaos commented on src/components/suggest/suggest_input.js:
```diff
@@ -0,0 +1,137 @@
+      unsaved: {
+        icon: 'dot',
+        color: 'accent',
+        tooltip: "You've made changes that haven't been saved yet.",
```
And this one could be more succinct too.
```suggestion
        tooltip: "Changes have not been saved.",
```

chandlerprall commented on src/components/suggest/suggest_item.js:
```diff
@@ -51,14 +58,22 @@ export const EuiSuggestItem = ({
   return (
-    <div className={classes} {...rest}>
+    <OuterElement
+      onClick={onClick ? onClick : undefined}
```
No need for the ternary here, `onClick={onClick}` is enough as it is already either itself or undefined

andreadelrio commented:
@chandlerprall thanks for your review. I've made changes following your suggestions.

cchaos commented on src-docs/src/views/suggest/global_filter_form.js:
```diff
@@ -75,26 +75,22 @@ export default class GlobalFilterForm extends Component {
-      customLabel: '',
-    };
-  }
+  state = {
```
@chandlerprall Should all the `state`s have been pulled out of the constructor?

chandlerprall replied:
Yep! This change is great to see

snide commented:
This will need a changelog update, but is now safe to merge. Congrats 🎉

snide commented:
@andreadelrio let's get this in 🚀

flaviolivolsi commented on src/components/suggest/suggest.js:
```diff
@@ -0,0 +1,81 @@
+
+    const suggestionList = suggestions.map((item, index) => (
+      <EuiSuggestItem
+        type={item.type}
```
`className` is not mapped here. Shall I submit a PR? @andreadelrio
